File: graffiti/src/main/java/com/ue/graffiti/widget/TextImageView.java
package com.ue.graffiti.widget;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.PointF;
import android.graphics.Typeface;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import com.ue.graffiti.R;
import com.ue.graffiti.event.BarSensorListener;
import com.ue.graffiti.model.Picture;
import com.ue.graffiti.model.Text;
import com.ue.graffiti.touch.TextImageTouch;
import com.ue.graffiti.util.TouchUtils;

public class TextImageView extends View {
    private Bitmap savedBitmap;
    private TextImageTouch touch;
    private PointF textPoint;//文字坐标
    private PointF centerPoint;//文字中心

    private int contentId;
    private Bitmap imageContent;

    private Paint drawTextPaint;
    private String textContent = "";

    private BarSensorListener mBarSensorListener;
    //0:text,1:image
    private int type;

    public TextImageView(Context context) {
        this(context, null);
    }

    public TextImageView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public TextImageView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        TypedArray ta = context.obtainStyledAttributes(attrs, R.styleable.TextImageView);
        type = ta.getInt(R.styleable.TextImageView_type, 0);
        if (type == 0) {
            //text
            drawTextPaint = new Paint();
            drawTextPaint.setTextSize(50);
            drawTextPaint.setTypeface(Typeface.DEFAULT_BOLD);
        } else {
            //image
            textPoint = new PointF();
            centerPoint = new PointF();
        }
        ta.recycle();
    }

    public void setBarSensorListener(BarSensorListener barSensorListener) {
        mBarSensorListener = barSensorListener;
    }

    public void setTextColor(int color) {
        drawTextPaint.setColor(color);
        invalidate();
    }

    public void setBitmap(Bitmap savedBitmap, int canvasWidth, int canvasHeight, int paintColor) {
        this.savedBitmap = Bitmap.createScaledBitmap(savedBitmap, canvasWidth, canvasHeight, true);

        touch = new TextImageTouch(true, canvasWidth, canvasHeight);
        if (mBarSensorListener != null) {
            touch.setBarSensorListener(mBarSensorListener);
        }

        drawTextPaint.setColor(paintColor);

        textPoint = new PointF();
        textPoint.set(touch.getTextPoint());
        centerPoint = new PointF();
        centerPoint.set(touch.getCenterPoint());
    }

    public void setBitmap(Bitmap savedBitmap, int canvasWidth, int canvasHeight) {
        this.savedBitmap = Bitmap.createScaledBitmap(savedBitmap, canvasWidth, canvasHeight, true);

        textPoint.set(canvasWidth / 2.5f, canvasHeight / 2.5f);
        centerPoint.set(textPoint);

        touch = new TextImageTouch(false, canvasWidth, canvasHeight);
        if (mBarSensorListener != null) {
            touch.setBarSensorListener(mBarSensorListener);
        }
    }

    public Text getText(int paintColor) {
        return new Text(textContent,
                touch.getDx(), touch.getDy(), touch.getScale(), touch.getDegree(),
                new PointF(touch.getCenterPoint().x, touch.getCenterPoint().y),
                new PointF(touch.getTextPoint().x, touch.getTextPoint().y), paintColor);
    }

    public Picture getPicture() {
        return new Picture(contentId,
                touch.getDx(), touch.getDy(), touch.getScale(), touch.getDegree(),
                new PointF(centerPoint.x, centerPoint.y),
                new PointF(textPoint.x, textPoint.y));
    }

    //触摸事件
    public boolean onTouchEvent(MotionEvent event) {
        touch.setCurPoint(new PointF(event.getX(0), event.getY(0)));
        touch.setSecPoint(event.getPointerCount() > 1 ? new PointF(event.getX(1), event.getY(1)) : new PointF(1, 1));

        int actionMasked = event.getActionMasked();
        switch (actionMasked) {
            // 第一只手指按下
            case MotionEvent.ACTION_DOWN:
                if (mBarSensorListener != null && mBarSensorListener.isTopToolbarVisible()) {
                    mBarSensorListener.closeTools();
                    touch.setDis(Float.MAX_VALUE);
                }
                touch.down1();
                break;
            case MotionEvent.ACTION_POINTER_DOWN:
                // 第二个手指按下
                touch.down2();
                break;
            case MotionEvent.ACTION_MOVE:
                touch.move();
                break;
            // 第一只手指抬起
            case MotionEvent.ACTION_UP:
                //第二只手抬起
            case MotionEvent.ACTION_POINTER_UP:
                touch.up();
                break;
        }
        invalidate();

        return true;
    }

    //重绘
    protected void onDraw(Canvas canvas) {
        canvas.drawBitmap(savedBitmap, 0, 0, new Paint());// 画其余图元
        canvas.translate(touch.getDx(), touch.getDy());
        canvas.scale(touch.getScale(), touch.getScale(), centerPoint.x, centerPoint.y);
        canvas.rotate(touch.getDegree(), centerPoint.x, centerPoint.y);

        if (type == 0) {
            //text
            canvas.drawText(textContent, textPoint.x, textPoint.y, drawTextPaint);
            return;
        }
        //image
        if (imageContent != null) {
            canvas.drawBitmap(imageContent, textPoint.x, textPoint.y, null);
        }
    }

    public void setTextContent(String textContent) {
        this.textContent = textContent;
    }

    public void setContentAndCenterPoint(int contentId) {
        this.contentId = contentId;
        TouchUtils.INSTANCE.ensureBitmapRecycled(imageContent);
        this.imageContent = BitmapFactory.decodeResource(getContext().getResources(), contentId);
        centerPoint.set(textPoint.x + imageContent.getWidth() / 2, textPoint.y + imageContent.getHeight() / 2);
    }

    public TextImageTouch getTouch() {
        return touch;
    }

    public Bitmap getImageContent() {
        return imageContent;
    }
}

File: graffiti/src/main/java/com/ue/graffiti/widget/TextImageView.kt
package com.ue.graffiti.widget

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import com.ue.graffiti.R
import com.ue.graffiti.event.BarSensorListener
import com.ue.graffiti.model.Picture
import com.ue.graffiti.model.Text
import com.ue.graffiti.touch.TextImageTouch
import com.ue.graffiti.util.ensureBitmapRecycled

class TextImageView : View {
    private var savedBitmap: Bitmap? = null
    lateinit var touch: TextImageTouch
    //文字坐标
    private lateinit var textPoint: PointF
    //文字中心
    private lateinit var centerPoint: PointF

    private var contentId = 0
    var imageContent: Bitmap? = null

    private lateinit var drawTextPaint: Paint
    private var textContent = ""

    var barSensorListener: BarSensorListener? = null
    //0:text,1:image
    private val type: Int

    val picture: Picture
        get() = Picture(contentId,
                touch.dx, touch.dy, touch.scale, touch.degree,
                PointF(centerPoint.x, centerPoint.y),
                PointF(textPoint.x, textPoint.y))

    constructor(context: Context) : this(context, null, 0)
    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        val ta = context.obtainStyledAttributes(attrs, R.styleable.gr_TextImageView)
        type = ta.getInt(R.styleable.gr_TextImageView_gr_type, 0)
        ta.recycle()
        if (type == 0) {
            //text
            drawTextPaint = Paint().apply {
                textSize = 50f
                typeface = Typeface.DEFAULT_BOLD
            }
        } else {
            //image
            textPoint = PointF()
            centerPoint = PointF()
        }
    }

    fun setTextColor(color: Int) {
        drawTextPaint.color = color
        invalidate()
    }

    fun setBitmap(savedBitmap: Bitmap, canvasWidth: Int, canvasHeight: Int, paintColor: Int) {
        this.savedBitmap = Bitmap.createScaledBitmap(savedBitmap, canvasWidth, canvasHeight, true)

        touch = TextImageTouch(true, canvasWidth, canvasHeight)
        touch.barSensorListener = barSensorListener

        drawTextPaint.color = paintColor

        textPoint = PointF()
        textPoint.set(touch.textPoint)
        centerPoint = PointF()
        centerPoint.set(touch.centerPoint)
    }

    fun setBitmap(savedBitmap: Bitmap, canvasWidth: Int, canvasHeight: Int) {
        this.savedBitmap = Bitmap.createScaledBitmap(savedBitmap, canvasWidth, canvasHeight, true)

        textPoint.set(canvasWidth / 2.5f, canvasHeight / 2.5f)
        centerPoint.set(textPoint)

        touch = TextImageTouch(false, canvasWidth, canvasHeight)
        touch.barSensorListener = barSensorListener
    }

    fun getText(paintColor: Int): Text {
        return Text(textContent,
                touch.dx, touch.dy, touch.scale, touch.degree,
                PointF(touch.centerPoint.x, touch.centerPoint.y),
                PointF(touch.textPoint.x, touch.textPoint.y), paintColor)
    }

    //触摸事件
    override fun onTouchEvent(event: MotionEvent): Boolean {
        touch.setCurrentPoint(PointF(event.getX(0), event.getY(0)))
        touch.setSecondPoint(if (event.pointerCount > 1) PointF(event.getX(1), event.getY(1)) else PointF(1f, 1f))

        val actionMasked = event.actionMasked
        when (actionMasked) {
        // 第一只手指按下
            MotionEvent.ACTION_DOWN -> {
                if (barSensorListener != null && barSensorListener!!.isTopToolbarVisible()) {
                    barSensorListener!!.closeTools()
                    touch.setDis(java.lang.Float.MAX_VALUE)
                }
                touch.down1()
            }
            MotionEvent.ACTION_POINTER_DOWN ->
                // 第二个手指按下
                touch.down2()
            MotionEvent.ACTION_MOVE -> touch.move()
        // 第一只手指抬起
            MotionEvent.ACTION_UP,
                //第二只手抬起
            MotionEvent.ACTION_POINTER_UP -> touch.up()
        }
        invalidate()

        return true
    }

    //重绘
    override fun onDraw(canvas: Canvas) {
        // 画其余图元
        canvas.drawBitmap(savedBitmap, 0f, 0f, Paint())
        canvas.translate(touch.dx, touch.dy)
        canvas.scale(touch.scale, touch.scale, centerPoint.x, centerPoint.y)
        canvas.rotate(touch.degree, centerPoint.x, centerPoint.y)

        if (type == 0) {
            //text
            canvas.drawText(textContent, textPoint.x, textPoint.y, drawTextPaint)
            return
        }
        if (imageContent != null) {
            //image
            canvas.drawBitmap(imageContent, textPoint.x, textPoint.y, null)
        }
    }

    fun setTextContent(textContent: String) {
        this.textContent = textContent
    }

    fun setContentAndCenterPoint(contentId: Int) {
        this.contentId = contentId
        ensureBitmapRecycled(imageContent)
        this.imageContent = BitmapFactory.decodeResource(context.resources, contentId)
        centerPoint.set(textPoint.x + imageContent!!.width / 2, textPoint.y + imageContent!!.height / 2)
    }
}